### PR TITLE
[Merged by Bors] - chore(GroupTheory/Perm/Cycle/Basic): Split

### PR DESCRIPTION
**Pull request by AntoineChambert-Loir**
The file Mathlib.GroupTheory.Perm.Cycle.Basic was too big and this PR splits it in several components:
* `Mathlib.GroupTheory.Perm.Cycle.Basic` contains everything related to a permutation being a cycle,
* `Mathlib.GroupTheory.Perm.Cycle.Factors` is about the cycles of a permutation and the decomposition of a permutation into disjoint cycles
* `Mathlib.GroupTheory.Perm.Closure` contains generation results for the permutation groups
* `Mathlib.GroupTheory.Perm.Finite` contains general results specific to permutation of finite types

I moved some results to `Mathlib.GroupTheory.Perm.Support`

I also moved some results from `Mathlib.GroupTheory.Perm.Sign` to `Mathlib.GroupTheory.Perm.Finite`

Some imports could be reduced, and the shake linter required a few adjustments in some other.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by YaelDillies — COMMENTED** [review]
Sorry, I'm really not convinced by this split. Please just ignore the linter in #9602 (you can edit the nolint entry). I will split the file later.

File: Mathlib/GroupTheory/Perm/Closure.lean [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+import Mathlib.GroupTheory.Perm.Cycle.Basic
+
+-- import Mathlib.Algebra.Module.BigOperators
+-- import Mathlib.Data.Finset.NoncommProd
+-- import Mathlib.Data.Fintype.Perm
+-- import Mathlib.Data.Int.ModEq
+-- import Mathlib.GroupTheory.Perm.List
+-- import Mathlib.GroupTheory.Perm.Sign
+-- import Mathlib.Logic.Equiv.Fintype
```
Comment (jcommelin) [inline Mathlib/GroupTheory/Perm/Closure.lean:15]: Please remove these.

**Review by jcommelin — APPROVED** [review]
@YaelDillies I don't understand what is so terrible about this split. I think it is a clear improvement over the status quo, and you're of course free to improve further in the future.

**Comment by riccardobrasca** [comment]
> @YaelDillies I don't understand what is so terrible about this split. I think it is a clear improvement over the status quo, and you're of course free to improve further in the future.

I agree that it's an improvement. I propose to merge it unless there a specific reason against it (also it risks to rot).

File: Mathlib/LinearAlgebra/Alternating/DomCoprod.lean [hunk 2]
```diff
@@ -5,7 +5,8 @@ Authors: Eric Wieser
 -/
 import Mathlib.LinearAlgebra.Alternating.Basic
 import Mathlib.LinearAlgebra.Multilinear.TensorProduct
-
+--
```
Comment (Ruben-VandeVelde) [inline Mathlib/LinearAlgebra/Alternating/DomCoprod.lean:8]:
```suggestion
```

File: Mathlib/GroupTheory/Perm/Sign.lean [hunk 3]
```diff
@@ -28,11 +32,13 @@ open Equiv Function Fintype Finset
...
 open BigOperators
 
-variable {α : Type u} {β : Type v}
+variable {α : Type u} [DecidableEq α] {β : Type v}
 
+/-
 -- An example on how to determine the order of an element of a finite group.
 example : orderOf (-1 : ℤˣ) = 2 :=
   orderOf_eq_prime (Int.units_sq _) (by decide)
+-/
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Perm/Sign.lean:41]: Should this be left commented out?

File: Mathlib/GroupTheory/Perm/Option.lean [hunk 4]
```diff
@@ -5,6 +5,9 @@ Authors: Eric Wieser
 -/
 import Mathlib.Data.Fintype.Perm
 import Mathlib.GroupTheory.Perm.Sign
+import Mathlib.Data.Fintype.Prod
+import Mathlib.Data.Fintype.Option
+--
 import Mathlib.Logic.Equiv.Option
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Perm/Option.lean:11]: Please sort and remove the `--`

File: Mathlib/LinearAlgebra/Alternating/DomCoprod.lean (hunk 2 above)
Comment (AntoineChambert-Loir) [reply Mathlib/LinearAlgebra/Alternating/DomCoprod.lean:8]: done

File: Mathlib/GroupTheory/Perm/Sign.lean (hunk 3 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Sign.lean:41]: The file does not import Mathlib.GroupTheory.OrderOfElement anymore, so the example does not compile.

File: Mathlib/GroupTheory/Perm/Option.lean (hunk 4 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Option.lean:11]: done

File: Mathlib/GroupTheory/Perm/Sign.lean (hunk 3 above)
Comment (jcommelin) [reply Mathlib/GroupTheory/Perm/Sign.lean]: @AntoineChambert-Loir Do you think it makes sense to move it to a place where it does typecheck? If so, please mvoe it.

Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Sign.lean]: I believe it was here because it has something to do with testing whether a permutation is even or odd… but the file where docs#orderOf is defined seems to be a good place.

Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Sign.lean]: finally, I put it in Mathlib/GroupTheory/SpecificGroups/Alternating (otherwise, I needed to add an import…)

File: Mathlib/GroupTheory/Perm/Option.lean (hunk 4 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Option.lean]: done

File: Mathlib/GroupTheory/Perm/Sign.lean [hunk 5]
```diff
@@ -13,26 +18,20 @@ import Mathlib.Data.Int.Order.Units
 /-!
 # Sign of a permutation
 
-The main definition of this file is `Equiv.Perm.sign`, associating a `ℤˣ` sign with a
-permutation.
+The main definition of this file is `Equiv.Perm.sign`,
+associating a `ℤˣ` sign with a permutation.
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/Perm/Sign.lean:22]:
```suggestion
The main definition of this file is `Equiv.Perm.sign`, associating a `ℤˣ` sign with a
permutation.
```

File: Mathlib/GroupTheory/Perm/Cycle/Factors.lean [hunk 6]
```diff
@@ -0,0 +1,694 @@
...
+
+import Mathlib.Algebra.Module.BigOperators
+import Mathlib.Data.Finset.NoncommProd
+import Mathlib.Data.Fintype.Perm
+import Mathlib.Data.Int.ModEq
+import Mathlib.GroupTheory.Perm.List
+import Mathlib.GroupTheory.Perm.Sign
+import Mathlib.Logic.Equiv.Fintype
+
+import Mathlib.GroupTheory.Perm.Cycle.Basic
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/Perm/Cycle/Factors.lean:15]:
```suggestion
import Mathlib.Data.Finset.NoncommProd
import Mathlib.Data.Int.ModEq
import Mathlib.GroupTheory.Perm.Cycle.Basic
import Mathlib.Logic.Equiv.Fintype
```

File: Mathlib/GroupTheory/Perm/Cycle/Factors.lean [hunk 7]
```diff
@@ -0,0 +1,694 @@
...
+
+/-!
+# Cycle factors of a permutation
+
+Let `β` be a `Fintype` and `f : Equiv.Perm β`.
+
+* `Equiv.Perm.cycleOf`: `f.cycleOf x` is the cycle of `f` that `x` belongs to.
+* `Equiv.Perm.cycleFactors`: `f.cycleFactors` is a list of disjoint cyclic permutations that
+  multiply to `f`.
+
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/Perm/Cycle/Factors.lean:27]: Sorry, this is unrelated to the split, but I've just noticed something is wrong with our generality
```suggestion

## TODO

Generalise `Equiv.Perm.cycleOf` to arbitrary types.
```

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean [hunk 8]
```diff
@@ -3,20 +3,19 @@ Copyright (c) 2019 Chris Hughes. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Chris Hughes, Yaël Dillies
 -/
-import Mathlib.Algebra.Module.BigOperators
-import Mathlib.Data.Finset.NoncommProd
+
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/Perm/Cycle/Basic.lean:6]:
```suggestion
import Mathlib.Algebra.Module.BigOperators
```

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean [hunk 9]
```diff
@@ -3,20 +3,19 @@ Copyright (c) 2019 Chris Hughes. All rights reserved.
...
 import Mathlib.Data.Fintype.Perm
-import Mathlib.Data.Int.ModEq
 import Mathlib.GroupTheory.Perm.List
-import Mathlib.GroupTheory.Perm.Sign
-import Mathlib.Logic.Equiv.Fintype
+import Mathlib.GroupTheory.Perm.Finite
+import Mathlib.Algebra.Module.BigOperators
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/Perm/Cycle/Basic.lean:10]:
```suggestion
```

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean [hunk 10]
```diff
@@ -60,13 +42,14 @@ open BigOperators
...
 
-namespace Equiv.Perm
 
 /-! ### `SameCycle` -/
 
 
 section SameCycle
 
+namespace Equiv.Perm
+
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/Perm/Cycle/Basic.lean:52]: What's going on here?

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean [hunk 11]
```diff
@@ -276,15 +259,18 @@ instance [Fintype α] [DecidableEq α] (f : Perm α) : DecidableRel (SameCycle f
...
 
 /-!
 ### `IsCycle`
 -/
 
-
 section IsCycle
 
+namespace Equiv.Perm
+
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/Perm/Cycle/Basic.lean:273]: and here

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean [hunk 12]
```diff
@@ -730,13 +716,78 @@ theorem IsCycle.isCycle_pow_pos_of_lt_prime_order [Finite β] {f : Perm β} (hf
...
+theorem _root_.Int.addRight_one_isCycle : (Equiv.addRight 1 : Perm ℤ).IsCycle :=
+  ⟨0, one_ne_zero, fun n _ => ⟨n, by simp⟩⟩
+#align int.add_right_one_is_cycle Int.addRight_one_isCycle
+
+end Int
+
+section Conjugation
+
+namespace Equiv.Perm
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/Perm/Cycle/Basic.lean:737]: and here (why did these two lemmas move?)

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean [hunk 13]
```diff
@@ -1907,6 +1066,7 @@ theorem _root_.Set.prod_self_eq_iUnion_perm (hf : f.IsCycleOn s) :
 
 end Set
 
+
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/Perm/Cycle/Basic.lean:1069]:
```suggestion
```

File: Mathlib/GroupTheory/Perm/Finite.lean [hunk 14]
```diff
@@ -0,0 +1,255 @@
+/-
+Copyright (c) 2018 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes
+-/
+import Mathlib.GroupTheory.Perm.Support
+import Mathlib.GroupTheory.OrderOfElement
+import Mathlib.Data.Finset.Fin
+import Mathlib.Data.Int.Order.Units
+import Mathlib.Logic.Equiv.Fintype
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/Perm/Finite.lean:10]:
```suggestion
import Mathlib.Data.Finset.Fin
import Mathlib.Data.Int.Order.Units
import Mathlib.GroupTheory.Perm.Support
import Mathlib.GroupTheory.OrderOfElement
import Mathlib.Logic.Equiv.Fintype
```

File: Mathlib/GroupTheory/Perm/Support.lean [hunk 15]
```diff
@@ -3,11 +3,12 @@ Copyright (c) 2018 Chris Hughes. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Chris Hughes, Aaron Anderson, Yakov Pechersky
 -/
-import Mathlib.Data.Finset.Card
-import Mathlib.Data.Fintype.Basic
+
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/Perm/Support.lean:6]:
```suggestion
```

File: Mathlib/GroupTheory/Perm/Sign.lean [hunk 16]
```diff
@@ -3,8 +3,13 @@ Copyright (c) 2018 Chris Hughes. All rights reserved.
...
+
 import Mathlib.GroupTheory.Perm.Support
-import Mathlib.GroupTheory.OrderOfElement
+import Mathlib.Tactic.NormNum.Ineq
+import Mathlib.Algebra.CharZero.Lemmas
+import Mathlib.Logic.Equiv.Fin
+import Mathlib.Data.Finset.Sort
+import Mathlib.GroupTheory.Subgroup.Finite
 import Mathlib.Data.Finset.Fin
 import Mathlib.Data.Int.Order.Units
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/Perm/Sign.lean:14]:
```suggestion
import Mathlib.Algebra.CharZero.Lemmas
import Mathlib.Data.Finset.Fin
import Mathlib.Data.Finset.Sort
import Mathlib.Data.Int.Order.Units
import Mathlib.GroupTheory.Perm.Support
import Mathlib.GroupTheory.Subgroup.Finite
import Mathlib.Logic.Equiv.Fin
import Mathlib.Tactic.NormNum.Ineq
```

File: Mathlib/GroupTheory/Perm/Sign.lean [hunk 17]
```diff
@@ -3,8 +3,13 @@ Copyright (c) 2018 Chris Hughes. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Chris Hughes
 -/
+
 import Mathlib.GroupTheory.Perm.Support
```
Comment (YaelDillies) [inline Mathlib/GroupTheory/Perm/Sign.lean:7]:
```suggestion
```

File: Mathlib/GroupTheory/Perm/Sign.lean
Comment (YaelDillies) [inline Mathlib/GroupTheory/Perm/Sign.lean:1]: Can you split off that split to its own PR? This one is getting out of hand

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean (hunk 10 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Cycle/Basic.lean:52]: What do you mean?

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean (hunk 12 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Cycle/Basic.lean:737]: Because I put all the `IsCycle`lemmas together, and then the `IsCycleOn`lemmas. Previously, everything was in an indefinite order.

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean (hunk 8 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Cycle/Basic.lean:6]: I don't understand what will be the effect of this suggestion.

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean (hunk 9 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Cycle/Basic.lean:10]: Idem.

File: Mathlib/GroupTheory/Perm/Finite.lean (hunk 14 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Finite.lean:10]: What is this change doing?

File: Mathlib/GroupTheory/Perm/Sign.lean (hunk 16 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Sign.lean:14]: You're adding an import?

File: Mathlib/GroupTheory/Perm/Sign.lean
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Sign.lean:1]: I would prefer not to.

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean (hunk 10 above)
Comment (YaelDillies) [reply Mathlib/GroupTheory/Perm/Cycle/Basic.lean:52]: Why did the namespace move inside the section? The namespace is supposed to be outside, with sections inside

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean (hunk 12 above)
Comment (YaelDillies) [reply Mathlib/GroupTheory/Perm/Cycle/Basic.lean:737]: I'm not sure "indefinite order" is very fair. Those lemmas were later because they talk about specific permutations rather than general ones.

If you really want to move those lemmas, you should keep the namespacing untouched and just add `_root_.Int` to their names
```suggestion

theorem _root_.Int.addLeft_one_isCycle : (Equiv.addLeft 1 : Perm ℤ).IsCycle :=
  ⟨0, one_ne_zero, fun n _ => ⟨n, by simp⟩⟩
#align int.add_left_one_is_cycle Int.addLeft_one_isCycle

theorem _root_.Int.addRight_one_isCycle : (Equiv.addRight 1 : Perm ℤ).IsCycle :=
  ⟨0, one_ne_zero, fun n _ => ⟨n, by simp⟩⟩
#align int.add_right_one_is_cycle Int.addRight_one_isCycle

section Conjugation
```

File: Mathlib/GroupTheory/Perm/Sign.lean (hunk 16 above)
Comment (YaelDillies) [reply Mathlib/GroupTheory/Perm/Sign.lean:14]: No, it's already there but Github doesn't let me do suggestions over deleted lines.

File: Mathlib/GroupTheory/Perm/Sign.lean (hunk 5 above)
Comment (jcommelin) [reply Mathlib/GroupTheory/Perm/Sign.lean:22]: I'm personally a fan of semantic linebreaks (https://sembr.org/).

We don't have a policy in mathlib concerning sembr or justifying text. I would leave this as it is.

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean (hunk 8 above)
Comment (YaelDillies) [reply Mathlib/GroupTheory/Perm/Cycle/Basic.lean:6]: I am ordering alphabetically the imports

File: Mathlib/GroupTheory/Perm/Finite.lean (hunk 14 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Finite.lean:10]: In fact, no…

Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Finite.lean:10]: sorted.

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean (hunk 10 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Cycle/Basic.lean:52]: Does closing a namespace close all introduction of variables?
It is not so clear in the documentation.

Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Cycle/Basic.lean:52]: Apparently, it does not. These `section`had the intent of isolating variables introduced here and there.

Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Cycle/Basic.lean:52]: I did however made the change.

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean (hunk 12 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Cycle/Basic.lean:737]: By “indefinite”, I meant that the small sections were alternating between stuff about IsCycle and IsCycleOn.  I just put all IsCycle lemmas in one place, and all IsCycleOn lemmas in another place.

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean (hunk 11 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Cycle/Basic.lean:273]: I made the change.

File: Mathlib/GroupTheory/Perm/Sign.lean (hunk 5 above)
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Sign.lean:22]: (That's not the subject here, but I am used to 80 chars terminals, and I find these overflow to 100 chars very disturbing, I get many lines with few symbols, hence less information on my screen.)

File: Mathlib/GroupTheory/Perm/Sign.lean
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Perm/Sign.lean:1]: Reverting the change to Sign would mean finding what to move and rechecking imports. The handful of lemmas I moved were about something else than Sign — just needed for that.
IIRC, some files that used `Sign` could stop using it.

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge
